fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingScrubDangling.java

```diff
@@ -0,0 +1,226 @@
+import com.apple.foundationdb.record.RecordMetaData;
+import com.apple.foundationdb.record.RecordMetaDataProvider;
+import com.apple.foundationdb.record.ScanProperties;
+import com.apple.foundationdb.record.TupleRange;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.metadata.IndexTypes;
+import com.apple.foundationdb.record.metadata.MetaDataException;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.protobuf.Message;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Manifesto:
+ *  Scrub a readable index to validate its consistency. Repair when needed.
+ *  Keep the type stamp simple, to allow parallel scrubbing.
+ *  Support two scrubbers - detect missing and detect dangling index entries. Set by the policy.
+ */
+@API(API.Status.INTERNAL)
+public class IndexingScrubDangling extends IndexingBase {
+    @Nonnull private static final Logger LOGGER = LoggerFactory.getLogger(IndexingScrubMissing.class);
+    @Nonnull private static final IndexBuildProto.IndexBuildIndexingStamp myIndexingTypeStamp = compileIndexingTypeStamp();
+
+    @Nonnull private final OnlineScrubber.ScrubbingPolicy scrubbingPolicy;
+    private long scanCounter = 0;
+
+    public IndexingScrubDangling(@Nonnull final IndexingCommon common,
+                                 @Nonnull final OnlineIndexer.IndexingPolicy policy) {
+        super(common, policy);
+        if (!common.isScrubber()) {
+            throw new MetaDataException("invalid scrubbing policy");
+        }
+        this.scrubbingPolicy = (OnlineScrubber.ScrubbingPolicy) common.getScrubbingPolicy();
+    }
+
+    @Override
+    List<Object> indexingLogMessageKeyValues() {
+        return Arrays.asList(
+                LogMessageKeys.INDEXING_METHOD, "scrub repair missing"
+        );
+    }
+
+    @Nonnull
+    @Override
+    IndexBuildProto.IndexBuildIndexingStamp getIndexingTypeStamp(final FDBRecordStore store) {
+        return myIndexingTypeStamp;
+    }
+
+    @Nonnull
+    static IndexBuildProto.IndexBuildIndexingStamp compileIndexingTypeStamp() {
+        return
+                IndexBuildProto.IndexBuildIndexingStamp.newBuilder()
+                        .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.SCRUB_REPAIR)
+                        .build();
+    }
+
+    @Override
+    CompletableFuture<Void> buildIndexInternalAsync() {
+        return getRunner().runAsync(context -> openRecordStore(context)
```

Comment:
I guess I'll reiterate a comment from a previous round that I don't think we want this in a runner. It also looks like the store isn't used here, and the only I/O it might do before it calls `scrubIndex` is to get a subspace from the subspace provider. I don't think that should be strictly necessary (the `subspace` should be something that it can create from the provider when needed), but it seems off that it creates the store at least.

Reply:
Sure. 
Should I eliminate the `openRecordStore` from `IndexingByIndex` too?

Reply:
I'll leave IndexingByIndex for another time. Changing it will require its own test.